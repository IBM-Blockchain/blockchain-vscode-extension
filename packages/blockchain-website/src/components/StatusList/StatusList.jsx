import React, { Component } from 'react'; 
import IssuesList from '../IssuesList/IssuesList';
import Axios from 'axios';

const githubProject = 'IBM-Blockchain/blockchain-vscode-extension';

//Github URLs
const issuesStatusApiUrl = 'https://api.github.com/repos/'+githubProject+'/issues?labels=status'

class StatusList extends Component {

    constructor() {
        super();
        this.state = {
            issues: undefined,
            error: undefined
        }

        this.refreshIssues = this.refreshIssues.bind(this);
    }

    async componentDidMount(){
        await this.refreshIssues();
    }

    async refreshIssues() {

        const response = await Axios.get(issuesStatusApiUrl);
        if (response) {
            this.setState({
                issues : response.data
            });
        } else {
            this.setState({
                error : 'Cannot load incidents from Github, sorry.'
            });
        }
    }

    render() {
        if (this.state.issues || this.state.error) {
            return (
                <div>
                    <IssuesList 
                        key='issues' 
                        issues={this.state.issues}
                        error={this.state.error}
                        refreshIssues={this.refreshIssues}
                        title='Current Incidents'
                        issueLabel='status'>
                    </IssuesList>
                </div>
            );
        } else {
            return <></>;
        }
    }
}

export default StatusList;
