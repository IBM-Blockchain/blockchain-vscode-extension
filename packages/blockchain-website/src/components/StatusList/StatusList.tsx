import React, { Component } from 'react';
import IssuesList from '../IssuesList/IssuesList';
import Axios, { AxiosResponse } from 'axios';

const githubProject: string = 'IBM-Blockchain/blockchain-vscode-extension';

// Github URLs
const issuesStatusApiUrl: string = 'https://api.github.com/repos/' + githubProject + '/issues?labels=status';

interface StatusState {
    issues: any;
    error: string | undefined;
}

class StatusList extends Component<{}, StatusState> {

    constructor(props: Readonly<StatusState>) {
        super(props);
        this.state = {
            issues: undefined,
            error: undefined
        };

        this.refreshIssues = this.refreshIssues.bind(this);
    }

    async componentDidMount(): Promise<void> {
        await this.refreshIssues();
    }

    async refreshIssues(): Promise<void> {

        try {
            const response: AxiosResponse = await Axios.get(issuesStatusApiUrl);
            this.setState({
                issues : response.data
            });
        } catch (error) {
            this.setState({
                error : 'Cannot load incidents from Github, sorry.'
            });
        }
    }

    render(): JSX.Element {
        if (this.state.issues || this.state.error) {
            return (
                <div>
                    <IssuesList
                        key='issues'
                        issues={this.state.issues}
                        error={this.state.error}
                        title='Known current issues'
                        issueLabel='status'>
                    </IssuesList>
                </div>
            );
        } else {
            return <></>;
        }
    }
}

export default StatusList;
