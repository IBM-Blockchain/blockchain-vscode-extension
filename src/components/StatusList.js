import React, { Component } from 'react'; 
import IssuesList from './IssuesList';
import $ from "jquery";

var githubProject = 'IBM-Blockchain/blockchain-vscode-extension';

//Github URLs
var issuesStatusApiUrl = 'https://api.github.com/repos/'+githubProject+'/issues?labels=status'

class StatusList extends Component {

    constructor() {
        super();
        this.state = {
            issues: undefined,
            error: undefined,
            refreshing: false
        }

        this.refreshIssues = this.refreshIssues.bind(this);
    }

    componentDidMount(){
        this.refreshIssues()
        // this.checkerId = setInterval(this.refreshIssues,10000);
    }

    componentWillUnmount(){
        // clearInterval(this.refreshIssues);
    }

    refreshIssues(e){

        if (e !== undefined) {
            e.preventDefault();
        }

        this.setState({
            refreshing : true
        });

        var updateIssues = function(data){
            this.setState({
                issues : data
            });
            //we add a timeout to give some UI feedback on the loading even if it is instantaneous
            setTimeout(function(){
                this.setState({
                    refreshing : false
                })
            }.bind(this),200);
        }.bind(this)

        $.get({
            url : issuesStatusApiUrl,
            success : updateIssues,
            error : function(){
                //we add a timeout to give some UI feedback on the loading even if it is instantaneous
                setTimeout(function(){
                    this.setState({
                        refreshing : false
                    });
                }.bind(this),200);
                this.setState({
                    error : 'Cannot load incidents from Github, sorry.'
                });
            }.bind(this)
        });
    }

    getInitialState(){
        return {
            refreshing : false
        }
    }
    render() {
        if (this.state.issues || this.state.error) {
            return (
                <div>
                    <IssuesList 
                        key='issues' issues={this.state.issues}
                        refreshing={this.state.refreshing}
                        error={this.state.error}
                        refreshIssues={this.refreshIssues}
                        title='Latest Incidents'
                        issueLabel='status'>
                    </IssuesList>
                </div>
            );
        } else {
            return <></>;
        }
    }
}

export default StatusList;
