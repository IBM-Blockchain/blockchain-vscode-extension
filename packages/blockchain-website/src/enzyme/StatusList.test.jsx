import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import chai from 'chai';
import sinon from 'sinon';
import Axios from 'axios';
import sinonChai from 'sinon-chai';
import StatusList from '../components/StatusList/StatusList.jsx';
jest.useFakeTimers();
chai.should();
chai.use(sinonChai);

describe('StatusList', () => {
    let mySandBox;
    let axiosGetStub;
    let clock;

    beforeEach(async() => {
        mySandBox = sinon.createSandbox();
        axiosGetStub = mySandBox.stub(Axios, 'get');
        axiosGetStub.returns({data: [
        {
            url: 'incident1',
            title: 'someIncident'
        },
        {
            url: 'incident2',
            title: 'someOtherIncident'
        }]});
        clock = sinon.useFakeTimers(); 
    });

    afterEach(async () => {
        mySandBox.restore();
        clock.restore();
    });

    it('should render the expected snapshot', async() => {
        const component = renderer
            .create(<StatusList />)
            .toJSON(); 
        expect(component).toMatchSnapshot();
    });

    it('should get latest incidents with a successful api call', async() => {   
        const issues = [
            {
                url: 'incident1',
                title: 'someIncident'
            },
            {
                url: 'incident2',
                title: 'someOtherIncident'
            }];  
        const component = await mount(<StatusList/>);
        axiosGetStub.should.have.been.calledOnce;
        component.state().should.deep.equal({issues, error: undefined});
    });

    it('should show error if API call is unsuccessful', async() => {
        axiosGetStub.returns(undefined);
        const component = await mount(<StatusList/>);
        clock.tick(200);
        component.state().error.should.equal('Cannot load incidents from Github, sorry.');
        axiosGetStub.should.have.been.calledOnce;
    });
});