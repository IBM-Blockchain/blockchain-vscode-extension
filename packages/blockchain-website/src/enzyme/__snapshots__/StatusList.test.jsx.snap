// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusList should render the expected snapshot 1`] = `null`;
